working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,917 @@
+  ListSize<ListSize<Object>> ys3 = ys; // OK.
+  ys[0].front(); // OK.
+  ys3.front().front(); // OK.
+  ys as List<List<String>>; // `ys` is promoted, succeeds at run time.
+}
+```
+
+
+## Syntax
+
+The rule for `<extensionDeclaration>` in the grammar is replaced by the
+following:
+
+```ebnf
+<extensionDeclaration> ::=
+  'protected'? 'extension' ('type'? <typeIdentifier> <typeParameters>?)?
+      <extensionShowHidePart> 'on' <type> <interfaces>? '{'
+    (<metadata> <extensionMemberDefinition>)*
+  '}'
+
+<extensionShowHidePart> ::=
+  <extensionShowClause>? <extensionHideClause>?
+
+<extensionShowClause> ::= 'show' <extensionShowHideList>
+
+<extensionHideClause> ::= 'hide' <extensionShowHideList>
+
+<extensionShowHideList> ::=
+  <extensionShowHideElement> (',' <extensionShowHideElement>)*
+
+<extensionShowHideElement> ::=
+  <type> | <identifier> | 'operator' <operator>
+```
+
+*In the rule `<extensionShowHideElement>`, note that `<type>` derives
+`<typeIdentifier>`, which makes `<identifier>` nearly redundant. However,
+`<identifier>` is still needed because it includes some strings that cannot
+be the name of a type but can be the basename of a member, e.g., the
+built-in identifiers.*
+
+
+## Static Analysis
+
+Assume that _E_ is an extension declaration of the following form:
+
+```dart
+extension Ext<X1 extends B1, .. Xk extends Bk> on T {
+  ... // Members
+}
+```
+
+*Note that `extension type` declarations and other enhanced forms are
+discussed in later sections. The properties specified here are valid for
+those forms as well, except for the differences which are specified
+explicitly.*
+
+It is then allowed to use `Ext<S1, .. Sk>` as a type: It can occur as the
+declared type of a variable or parameter, as the return type of a function
+or getter, as a type argument in a type, or as the on-type of an extension.
+
+*In particular, it is allowed to create a new instance where one or more
+extension types occur as type arguments.*
+
+When `k` is zero, `Ext<S1, .. Sk>` simply stands for `Ext`, a non-generic
+extension. When `k` is greater than zero, a raw occurrence `Ext` is treated
+like a raw type: Instantiation to bound is used to obtain the omitted type
+arguments.
+
+We say that the static type of said variable, parameter, etc. _is the
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is treated as
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>>(v).m(args)</code>
+where `v` is a fresh variable whose declared type
+is the on-type corresponding to `Ext<S1, .. Sm>`, and similarly for
+instance getters and operators. This rule also applies when a member access
+implicitly has the receiver `this`, and the static type of `this` is an
+extension type (*which can only occur in an extension type member
+declaration*).
```

Comment:
The situation where `this` can have an extension type as type is when the `on` type is an extension type:
```dart
extension type E1 on int {
  bool get isEven => false; // It's only half wrong!
}
extension type E2 on E1 {
  foo() {
    // No `isEven` in scope.
    // Found `isEven` on `E1`, use that.
    print(isEven); // So equivalent to `this.isEven` which is `E1(this as int).isEven`
  }
}
main() {
  E2(E1(3)).foo(); // prints false
}
```

Reply:
True. Added a few words to clarify that.